### PR TITLE
336-api-digital-shadow

### PR DESCRIPTION
**Pull request by larswitte**
Solves issue #336

### CODE REVIEW

**Comment by larswitte** [comment]
> **Current State**
> At the moment data from the `uniqueProductTable` gets enriched with data from the `productTagTable` and `productTagStringTable`.
> 
> **Problem**
> Additionally I would like to include the inheritance information. To be specific, I would like to include the AID's of parents. To get the AID we have to connect the `uniqueProductTable` with the `productInheritanceTable` and back:
> 
> `uniqueProductTable`(unique products filtered by asset and timeframe) <----------_[connected via **child UID** to find the parent UID]_-----------> `productInheritanceTable`<----------_[connected via **parent UID** to find the parent AID]_ ------------> `uniqueProductTable'
> 
> Down below is my first idea on how the sql query might look like, but I don't know, if I'm on the right track. Also I don't know, how to efficiently test different SQL queries (next I will try DataGrip with example Data).
> 
> ```sql
> 	SELECT u.uniqueProductID, u.uniqueProductAlternativeID, u.begin_timestamp_ms, u.end_timestamp_ms, u.product_id, u.is_scrap, m.uniqueProductAlternativeID
> 	FROM uniqueProductTable u, productInheritanceTable p, uniqueProductTable m 
> 		INNER JOIN p ON u.uniqueProductID = p.child_uid
> 		INNER JOIN m ON p.parent_uid = m.uniqueProductID
> 	WHERE u.asset_id = $1 
> 		AND (u.begin_timestamp_ms BETWEEN $2 AND $3 OR u.end_timestamp_ms BETWEEN $2 AND $3) 
> 		OR (u.begin_timestamp_ms < $2 AND u.end_timestamp_ms > $3) 
> 	ORDER BY u.uniqueProductID ASC;
> ```

After talking with @Gornoka about this issue, I changed the sql query (deleted unnecessary variables and changed the alias names).
Additionally I also query `product_name` from the `productTable` now:
```sql
SELECT unProdTab.uniqueProductID, unProdTab.begin_timestamp_ms, prodTab.product_name, unProdTabForAID.uniqueProductAlternativeID
	FROM uniqueProductTable unProdTab, productInheritanceTable prodInher, uniqueProductTable unProdTabForAID, productTable prodTab
		INNER JOIN prodInher ON unProdTab.uniqueProductID = prodInher.child_uid
		INNER JOIN unProdTabForAID ON prodInher.parent_uid = unProdTabForAID.uniqueProductID
		INNER JOIN prodTab ON unProdTabForAID.product_id = prodTab.product_id
	WHERE unProdTab.asset_id = $1 
		AND (unProdTab.begin_timestamp_ms BETWEEN $2 AND $3 OR unProdTab.end_timestamp_ms BETWEEN $2 AND $3) 
		OR (unProdTab.begin_timestamp_ms < $2 AND unProdTab.end_timestamp_ms > $3) 
	ORDER BY unProdTab.uniqueProductID ASC;
```

**Comment by larswitte** [comment]
Accidentally merged with united-manufacturing-hub/main.
I reverted the merge, but in github some commits and files are still listed.

**Comment by larswitte** [comment]
First integration tests were successful.

File: deployment/factorycube-server/values.yaml [hunk 1]
```diff
@@ -5,7 +5,10 @@
 ### factoryinsight ###
 factoryinsight:
   enabled: true
-  image: unitedmanufacturinghub/factoryinsight
+  image:
```
Comment (JeremyTheocharis) [inline deployment/factorycube-server/values.yaml:8]: Please use the unitedmanufacturinghub repo

**Review by JeremyTheocharis — CHANGES_REQUESTED** [review]
Have no checked the actual programming logic, will do that in the future

**Comment by larswitte** [comment]
The get uniqueProducts API now also works, I tested it.

File: golang/cmd/factoryinsight/database.go [hunk 2]
```diff
@@ -1544,3 +1551,241 @@ func GetAssetID(parentSpan opentracing.Span, customerID string, location string,
...
+		INNER JOIN productInheritanceTable prodInher ON unProdTab.uniqueProductID = prodInher.child_uid
+		INNER JOIN uniqueProductTable unProdTabForAID ON prodInher.parent_uid = unProdTabForAID.uniqueProductID
+		INNER JOIN productTable prodTab ON unProdTabForAID.product_id = prodTab.product_id
+	WHERE unProdTab.asset_id = $1 
+		AND (unProdTab.begin_timestamp_ms BETWEEN $2 AND $3 OR unProdTab.end_timestamp_ms BETWEEN $2 AND $3) 
+		OR (unProdTab.begin_timestamp_ms < $2 AND unProdTab.end_timestamp_ms > $3) 
+	ORDER BY unProdTab.uniqueProductID ASC;`
+
+
+	rows, err := db.Query(sqlStatementData, assetID, from, to)
```
[the remark this answers is not in the document]
Comment (larswitte) [reply golang/cmd/factoryinsight/database.go:1610]: I don't know, quickly looking it up, I didn't find an aswer. To be honest, I think we have bigger optimizing potential in the function `SliceContatinsInt`.

File: deployment/factorycube-server/values.yaml (hunk 1 above)
Comment (larswitte) [reply deployment/factorycube-server/values.yaml:8]: Done

File: golang/cmd/factoryinsight/database.go [hunk 3]
```diff
@@ -1289,6 +1291,11 @@ func GetUniqueProducts(parentSpan opentracing.Span, customerID string, location
 		return
 	}
 
+	err = CheckOutputDimensions(data.Datapoints, data.ColumnNames)
```
[the remark this answers is not in the document]
Comment (larswitte) [reply golang/cmd/factoryinsight/database.go:1295]: Done

**Comment by larswitte** [comment]
The branch #461 should be ready to merge. Can you review it @JeremyTheocharis  please?

**Review by JeremyTheocharis — APPROVED** [review]

**Pull request merged**